pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -423,102 +427,117 @@ func (a Adapter) waitAndGetComponentPod(hideSpinner bool) (*corev1.Pod, error) {
-		command := pushDevfileCommands[i]
+	type CommandNames struct {
+		defaultName string
+		adapterName string
+	}
 
-		// Exec the devBuild command if buildRequired is true
-		if (command.Name == string(common.DefaultDevfileBuildCommand) || command.Name == a.devfileBuildCmd) && buildRequired {
-			glog.V(3).Infof("Executing devfile command %v", command.Name)
+	commandOrder := []CommandNames{}
 
-			for _, action := range command.Actions {
-				// Change to the workdir and execute the command
-				var cmdArr []string
-				if action.Workdir != nil {
-					cmdArr = []string{"/bin/sh", "-c", "cd " + *action.Workdir + " && " + *action.Command}
-				} else {
-					cmdArr = []string{"/bin/sh", "-c", *action.Command}
-				}
+	if runInit {
+		commandOrder = append(commandOrder, CommandNames{defaultName: string(common.DefaultDevfileInitCommand), adapterName: a.devfileInitCmd})
+	}
+	commandOrder = append(
+		commandOrder,
+		CommandNames{defaultName: string(common.DefaultDevfileBuildCommand), adapterName: a.devfileBuildCmd},
+		CommandNames{defaultName: string(common.DefaultDevfileRunCommand), adapterName: a.devfileRunCmd},
+	)
 
-				if show {
-					s = log.SpinnerNoSpin("Executing " + command.Name + " command " + fmt.Sprintf("%q", *action.Command))
+	for i, currentCommand := range commandOrder {
+		for _, command := range pushDevfileCommands {
+			if command.Name == currentCommand.defaultName || command.Name == currentCommand.adapterName {
+				if i < len(commandOrder)-1 {
+					// Any exec command such as "Init" and "Build"
+					err := a.executeDevfileCommand(command, show, podName)
+					if err != nil {
+						return err
+					}
 				} else {
-					s = log.Spinner("Executing " + command.Name + " command " + fmt.Sprintf("%q", *action.Command))
-				}
-
-				defer s.End(false)
-
-				err = exec.ExecuteCommand(&a.Client, podName, *action.Component, cmdArr, show)
-				if err != nil {
-					s.End(false)
-					return err
+					// Last command is "Run"
+					glog.V(3).Infof("Executing devfile command %v", command.Name)
+
+					for _, action := range command.Actions {
+
+						// Check if the devfile run component containers have supervisord as the entrypoint.
+						// Start the supervisord if the odo component does not exist
+						if !componentExists {
+							err = a.InitRunContainerSupervisord(*action.Component, podName, containers)
+							if err != nil {
+								return
+							}
+						}
+
+						// Exec the supervisord ctl stop and start for the devrun program
+						type devRunExecutable struct {
+							command []string
+						}
+						devRunExecs := []devRunExecutable{
+							{
+								command: []string{common.SupervisordBinaryPath, "ctl", "stop", "all"},
+							},
+							{
+								command: []string{common.SupervisordBinaryPath, "ctl", "start", string(common.DefaultDevfileRunCommand)},
+							},
+						}
+
+						s = log.Spinner("Executing " + command.Name + " command " + fmt.Sprintf("%q", *action.Command))
```

Comment:
Instead of `+` each string. Why not put the entire thing in a `fmt.Sprintf`?

Ex: `log.Spinner(fmt.Sprintf("Executing %s %s, etc, etc.", command.Name, *action.Command),` etc.